waku/v2/node/peer_manager/peer_manager.nim

```diff
@@ -235,6 +235,31 @@ proc loadFromStorage(pm: PeerManager) =
   else:
     debug "successfully queried peer storage"
 
+proc canBeConnected*(pm: PeerManager,
+                     peerId: PeerId): bool =
+  # Returns if we can try to connect to this peer, based on past failed attempts
+  # It uses an exponential backoff. Each connection attempt makes us
+  # wait more before trying again.
+  let failedAttempts = pm.peerStore[NumberFailedConnBook][peerId]
+
+  # if it never errored, we can try to connect
+  if failedAttempts == 0:
+    return true
+
+  # if there are too many failed attemps, do not reconnect 
+  # the check for 14 comes from overflow described in https://github.com/waku-org/nwaku/issues/1665)
+  if failedAttempts >= pm.maxFailedAttempts or failedAttempts > 14:
+    return false
+
+  # If it errored we wait an exponential backoff from last connection
+  # the more failed attemps, the greater the backoff since last attempt
+  let now = Moment.init(getTime().toUnix, Second)
+  let lastFailed = pm.peerStore[LastFailedConnBook][peerId]
+  let backoff = chronos.seconds(pm.initialBackoffInSec*(pm.backoffFactor^(failedAttempts-1)))
```

Comment:
Hmm.. can we abstract this into a function and fuzz it in the test suite? Will help catch these errors in the futures too

Reply:
not against fuzzing but perhaps overkill? related to my other comment, perhaps we can just assert that the backoff values makes sense when we init the peermanager with initialBackoffInSec and backoffFactor. just run the math in there to ensure values are withing some reasonable limits.

Reply:
that works too, just something we should consider next time :)